### PR TITLE
plugin: Unify reserve and unreserve logic

### PR DESCRIPTION
**Pull request by sharnoff**
We had the same thing implemented in a few places, with some TODOs to unify them eventually - so, here it is.

Summary of changes:

- Merge all handling of pod/VM "start" / "reserve" logic to go through `(*AutoscaleEnforcer).reserveResources()`
  - Combined `handlePodStarted`/`handleVMStarted` -> `handleStarted`
- Merge all handling of pod/VM "deletion" / "unreserve" logic to go through `(*AutoscaleEnforcer).unreserveResources()`
- Pass around `*corev1.Pod` objects 

TODO:

- [x] Add comments explaining new functions and their parameters
- [x] Add comments on substantially modified functions that didn't already have them (like `handlePodStarted`)

### CODE REVIEW

**Comment by sharnoff** [comment]
Goal: include this in the next release, after #653. That change was relatively mechanical, this is much higher risk. Want to evenly distribute risk across releases.

File: pkg/plugin/state.go [hunk 1]
```diff
@@ -569,98 +569,205 @@ func (e *AutoscaleEnforcer) handleNodeDeletion(logger *zap.Logger, nodeName stri
...
+	logFields, kind, migrating, verdict := e.unreserveResources(logger, podName)
+
+	logger.With(logFields...).Info(
+		fmt.Sprintf("Deleted %s Pod", kind),
+		zap.Bool("migrating", migrating),
+		zap.Object("verdict", verdict),
+	)
+}
+
+func (e *AutoscaleEnforcer) unreserveResources(
```
Comment (shayanh) [inline pkg/plugin/state.go:767]: As an outsider to me reserve and unreserve methods are symmetric operations. I don't get why here their inputs and outputs are so different.

File: pkg/plugin/state.go [hunk 2]
```diff
@@ -569,98 +569,205 @@ func (e *AutoscaleEnforcer) handleNodeDeletion(logger *zap.Logger, nodeName stri
...
+	logger *zap.Logger,
+	podName util.NamespacedName,
+) (_ []zap.Field, kind string, migrating bool, _ verdictSet) {
```
Comment (shayanh) [inline pkg/plugin/state.go:770]: To me having partially named return arguments looks slightly ugly. I'd rather suggest naming all or none.

File: pkg/plugin/state.go [hunk 3]
```diff
@@ -569,98 +569,205 @@ func (e *AutoscaleEnforcer) handleNodeDeletion(logger *zap.Logger, nodeName stri
...
+	_, _, _ = e.reserveResources(context.TODO(), logger, pod, "Pod started", false)
+}
+
+func (e *AutoscaleEnforcer) reserveResources(
+	ctx context.Context,
+	logger *zap.Logger,
+	pod *corev1.Pod,
+	action string,
+	allowDeny bool,
+) (ok bool, _ *verdictSet, _ error) {
```
Comment (shayanh) [inline pkg/plugin/state.go:595]: Why we return both an error and a boolean here?

File: pkg/plugin/state.go (hunk 2 above)
Comment (shayanh) [reply pkg/plugin/state.go:770]: Returning log fields here looks sorta hacky.

File: pkg/plugin/state.go (hunk 1 above)
Comment (sharnoff) [reply pkg/plugin/state.go:791]: Added some comments in f4b9d26cd4a19be1baec6fdf0899f04c79d2b9ac — wdyt?

File: pkg/plugin/state.go (hunk 3 above)
Comment (sharnoff) [reply pkg/plugin/state.go:610]: Elaborated in a comment in f4b9d26cd4a19be1baec6fdf0899f04c79d2b9ac — wdyt?

File: pkg/plugin/state.go (hunk 2 above)
Comment (sharnoff) [reply pkg/plugin/state.go:794]: I'd left a comment on the function. I agree parts of this are a little ugly/hacky. I don't see a cleaner solution tbh, although maybe I'm missing something.

wdyt? 

(In terms of named/unnamed returns, IMO having a mix allows documenting some fields while making it clear that you shouldn't use a naked return. Here, I think it's necessary to name `kind` and `migrating`.)

Comment (shayanh) [reply pkg/plugin/state.go:794]: Sounds good.

**Review by shayanh — APPROVED** [review]

**Pull request merged**